File: src/lib/mess_coder.c
/*
 * file:        mess_coder.c
 * author:      VasiliyMatlab
 * version:     1.0
 * date:        27.04.2023
 * copyright:   Vasiliy (c) 2023
 */

#include <stdio.h>

#include "mess_coder.h"

/**
 * \brief Кодирование данных
 * 
 * \param[out] out Указатель на закодированные данные
 * \param[in] size_out Ограничение по размеру на закодированные данные
 * \param[in] in Указатель на входные данные
 * \param[in] size_in Размер входных данных
 * \return Размер закодированных данных;
 * в случае ошибки - отрицательный код
 */
static int messcoder_encode(void *out, uint32_t size_out,
						 	const void *in, uint32_t size_in) {
	int rc;
	uint32_t idx_out = 0;
	uint8_t *istream = (uint8_t *) in;
	uint8_t *ostream = (uint8_t *) out;
	
	if (size_out == 0)
		return 0;
	
	// Добавляем байт начала
	ostream[idx_out++] = MESS_CODER_START_B;
	
	// Начинаем замену старт/стоп байтов; при этом в буфере должно 
	// оставаться достаточно места для вставки 2-х байт (замены)
	// или добавления байта окончания
	for (uint32_t idx_in = 0; (idx_in < size_in) && (idx_out < (size_out-1)); idx_in++) {
		switch (istream[idx_in]) {
		// Если встречаем символ начала посылки, то кодируем его
		case MESS_CODER_START_B:
			ostream[idx_out++] = MESS_CODER_ENC_START;		// спец символ
			ostream[idx_out++] = MESS_CODER_ENC_START_B;	// код начала посылки
			break;
		
		// Если встречаем спец символ, то кодируем его
		case MESS_CODER_ENC_START:
			ostream[idx_out++] = MESS_CODER_ENC_START;		// спец символ
			ostream[idx_out++] = MESS_CODER_ENC_DATA_B;		// код спец символа
			break;
		
		// Если встречаем символ конца посылки, то кодируем его
		case MESS_CODER_END_B:
			ostream[idx_out++] = MESS_CODER_ENC_START;		// спец символ
			ostream[idx_out++] = MESS_CODER_ENC_END_B;		// код конца посылки
			break;

		// Иначе оставляем символ как есть (без кодирования)
		default:
			ostream[idx_out++] = istream[idx_in];
			break;
		}
	}	
	
	// Добавляем байт окончания
	if (idx_out < size_out) {
		ostream[idx_out++] = MESS_CODER_END_B;
		rc = (int) idx_out;
	} else {
		// Переполнение выходного буфера
		rc = MESS_CODER_RC_OVERFLOW;
	}
	
	return rc;
}

/**
 * \brief Декодирование данных
 * 
 * \param[out] out Указатель на декодированные данные
 * \param[in] size_out Ограничение по размеру на декодированные данные
 * \param[in] in Указатель на входные данные
 * \param[in] size_in Размер входных данных
 * \return Размер декодированных данных;
 * в случае ошибки - отрицательный код
 */
static int messcoder_decode(void *out, uint32_t size_out,
			 			 	const void *in, uint32_t size_in) {
	int rc;
	uint32_t idx_in;
	uint32_t idx_out;
	uint8_t *istream = (uint8_t *) in;
	uint8_t *ostream = (uint8_t *) out;
	
	if(size_in == 0)
		return 0;
	
	// Ищем байт начала потока
	for (idx_in = 0; idx_in < size_in; idx_in++) {
		if (istream[idx_in] == MESS_CODER_START_B) {
			break;
		}
	}
	
	if (idx_in == size_in) {
		return MESS_CODER_RC_NO_START;
	}
	
	// Начинаем поиск последовательностей кодов и замену на исходные байты
	for (idx_out = 0; (idx_in < (size_in-1)) && (idx_out < size_out); idx_in++) {
		switch (istream[idx_in]) {
		// Нашли новое начало посылки;
		// сбрасываем счетчик выходного буфера и начинаем
		// писать заново
		case MESS_CODER_START_B:
			idx_out = 0;
			break;
		
		// Нашли конец посылки
		case MESS_CODER_END_B:
			goto check_end_byte;
		
		// Нашли байт начала кодовой последовательности;
		// расшифровываем следующий за ним байт для
		// восстановления исходной комбинации
		case MESS_CODER_ENC_START:
			switch (istream[idx_in + 1]) {
			// Следующий байт - код совпадения с началом посылки
			case MESS_CODER_ENC_START_B:
				ostream[idx_out++] = MESS_CODER_START_B;
				idx_in++;
				break;
			
			// Следующий байт - код совпадения со спец символом
			case MESS_CODER_ENC_DATA_B:
				ostream[idx_out++] = MESS_CODER_ENC_START;
				idx_in++;
				break;
			
			// Следующий байт - код совпадения с концом посылки
			case MESS_CODER_ENC_END_B:
				ostream[idx_out++] = MESS_CODER_END_B;
				idx_in++;
				break;
			
			// Неизвестная кодовая последовательность в
			// закодированном потоке (если мы это получили,
			// то это означает, что посылка была искажена
			// или кодировщик отправки потока отработал
			// неправильно)
			default:
				fprintf(stderr, "Error: MESS_CODER: invalid code 0x%2X\r\n",
						istream[idx_in+1]);
				return MESS_CODER_RC_DECERR;
			}
			break;
		
		// Нашли не закодированный байт
		default:
			ostream[idx_out++] = istream[idx_in];
			break;
		}
	}

check_end_byte:
	// Проверяем последний символ; поток должен завершаться
	// символом конца, иначе - ошибка
	if (istream[idx_in] != MESS_CODER_END_B) {
		if (idx_out >= size_out) {
			// Переполнение выходного буфера
			fprintf(stderr, "Error: MESS_CODER: output buffer overflow %u (avaliable %u)\r\n",
					idx_out+1, size_out);
			rc = MESS_CODER_RC_OVERFLOW;
		} else {
			// Нет символа окончания посылки
			rc = MESS_CODER_RC_NO_END;
		}
	} else {
		rc = (int) idx_out;
	}

	return rc;
}

// Преобразование блока данных в поток для передачи по последовательному интерфейсу
int messcoder_to_serial(void *out, uint32_t size_out,
		    		 	const void *in, uint32_t size_in) {
	if (!in || !size_in) {
		return MESS_CODER_RC_ERROR;
	}

	if (!out) {
		return MESS_CODER_RC_ERROR;
	}
	
	return messcoder_encode(out, size_out, in, size_in);
}

// Преобразование потока данных последовательного интерфейса в блок данных
int messcoder_from_serial(void *out, uint32_t size_out,
					   	  const void *in, uint32_t size_in) {
	if (!in || !size_in) {
		return MESS_CODER_RC_ERROR;
	}

	if (!out) {
		return MESS_CODER_RC_ERROR;
	}

	return messcoder_decode(out, size_out, in, size_in);
}

// Рассчитывание размера выходного буфера
int messcoder_comp_enc_size(const void *in, uint32_t size_in) {
	const uint8_t *istream = (const uint8_t *) in;
	uint32_t ostream_size = 2;
	
	for (uint32_t idx_in = 0; idx_in < size_in; idx_in++) {
		switch(istream[idx_in]) {
		case MESS_CODER_START_B:
		case MESS_CODER_END_B:
		case MESS_CODER_ENC_START:
			ostream_size += 2;
			break;
		default:
			ostream_size++;
			break;
		}
	}

	return (int) ostream_size;
}

File: python/messcoder.py
#!/usr/bin/env python3

"""messcoder.py: Message Coder module"""

__author__ = "VasiliyMatlab"
__copyright__ = "Vasiliy (c) 2023"


from enum import IntEnum

class Defines(IntEnum):
    MESS_CODER_START_B      = 0xAB  # Символ начала посылки (в закодированном потоке)
    MESS_CODER_END_B        = 0xCD  # Символ конца посылки (в закодированном потоке)

    MESS_CODER_ENC_START    = 0x1E  # Признак начала спец последовательности, после которого идет код спец последовательности (в закодированном потоке)
    MESS_CODER_ENC_START_B  = 0x01  # Код совпадения с началом посылки (в закодированном потоке)
    MESS_CODER_ENC_DATA_B   = 0x02  # Код совпадения с началом спец последовательности (в закодированном потоке)
    MESS_CODER_ENC_END_B    = 0x03  # Код совпадения с концом посылки (в закодированном потоке)

    MESS_CODER_RC_ERROR	    = -1    # Общая ошибка
    MESS_CODER_RC_NO_START  = -21   # Символ начала посылки не найден
    MESS_CODER_RC_NO_END    = -22   # Символ конца посылки не найден
    MESS_CODER_RC_OVERFLOW  = -23   # Нехватка места в выходном буфере
    MESS_CODER_RC_DECERR    = -24   # Ошибка декодирования ключевой последовательности


## \brief Кодирование данных
#
# \param[in] inp Список входных данных
# \return Список декодированных данных
def __encode(inp: list) -> list[int]:
    out = list()

    # Добавляем байт начала
    out.append(int(Defines.MESS_CODER_START_B))

    # Начинаем замену старт/стоп байтов
    for elem in inp:
        match elem:
            # Если встречаем символ начала посылки, то кодируем его
            case int(Defines.MESS_CODER_START_B):
                out.append(int(Defines.MESS_CODER_ENC_START))   # спец символ
                out.append(int(Defines.MESS_CODER_ENC_START_B)) # код начала посылки

            # Если встречаем спец символ, то кодируем его
            case int(Defines.MESS_CODER_ENC_START):
                out.append(int(Defines.MESS_CODER_ENC_START))   # спец символ
                out.append(int(Defines.MESS_CODER_ENC_DATA_B))  # код спец символа

            # Если встречаем символ конца посылки, то кодируем его
            case int(Defines.MESS_CODER_END_B):
                out.append(int(Defines.MESS_CODER_ENC_START))   # спец символ
                out.append(int(Defines.MESS_CODER_ENC_END_B))   # код конца посылки

            # Иначе оставляем символ как есть (без кодирования)
            case _:
                out.append(elem)

    # Добавляем байт окончания
    out.append(int(Defines.MESS_CODER_END_B))

    return out


## \brief Декодирование данных
#
# \param[in] inp Список входных данных
# \return Кортеж: код возврата и список декодированных данных
def __decode(inp: list) -> tuple[int, list[int]]:
    # Ищем байт начала потока
    idx = 0
    while idx < len(inp):
        if inp[idx] == int(Defines.MESS_CODER_START_B):
            break
        idx += 1
    else:
        return (int(Defines.MESS_CODER_RC_NO_START), list())
    
    # Начинаем поиск последовательностей кодов и замену на исходные байты
    out = list()
    try:
        while idx < (len(inp) - 1):
            match inp[idx]:
                # Нашли новое начало посылки;
                # выходной буфер и начинаем писать заново
                case int(Defines.MESS_CODER_START_B):
                    out.clear()

                # Нашли конец посылки
                case int(Defines.MESS_CODER_END_B):
                    raise StopIteration
                
                # Нашли байт начала кодовой последовательности;
                # расшифровываем следующий за ним байт для
                # восстановления исходной комбинации
                case int(Defines.MESS_CODER_ENC_START):
                    match inp[idx+1]:
                        # Следующий байт - код совпадения с началом посылки
                        case int(Defines.MESS_CODER_ENC_START_B):
                            out.append(int(Defines.MESS_CODER_START_B))
                            idx += 1
                        # Следующий байт - код совпадения со спец символом
                        case int(Defines.MESS_CODER_ENC_DATA_B):
                            out.append(int(Defines.MESS_CODER_ENC_START))
                            idx += 1
                        # Следующий байт - код совпадения с концом посылки
                        case int(Defines.MESS_CODER_ENC_END_B):
                            out.append(int(Defines.MESS_CODER_END_B))
                            idx += 1
                        # Неизвестная кодовая последовательность в
                        # закодированном потоке (если мы это получили,
                        # то это означает, что посылка была искажена
                        # или кодировщик отправки потока отработал
                        # неправильно)
                        case _:
                            print(f"Error: MESS_CODER: invalid code 0x{hex(inp[idx+1])}")
                            return (int(Defines.MESS_CODER_RC_DECERR), list())
                        
                # Нашли не закодированный байт
                case _:
                    out.append(inp[idx])

            idx += 1
    except StopIteration:
        pass

    # Проверяем последний символ; поток должен завершаться
	# символом конца, иначе - ошибка
    if inp[-1] != int(Defines.MESS_CODER_END_B):
        return (int(Defines.MESS_CODER_RC_NO_END), out)
    
    return (0, out)


## \brief Функция, преобразующая блок данных
#  в поток для передачи по последовательному интерфейсу;
#  добавляет символы начала и конца посылки
#
# \param[in] inp Список входных данных
# \return Кортеж: код возврата и список декодированных данных
def to_serial(inp: list) -> tuple[int, list[int]]:
    if not inp:
        return (int(Defines.MESS_CODER_RC_ERROR), list())
    return (0, __encode(inp))


## \brief Функция, преобразующая поток данных из последовательного интерфейса
#  в блок данных; убирает символы начала и конца посылки
#
# \param[in] inp Список входных данных
# \return Кортеж: код возврата и список декодированных данных
def from_serial(inp: list) -> tuple[int, list[int]]:
    if not inp:
        return (int(Defines.MESS_CODER_RC_ERROR), list())
    return __decode(inp)


## \brief Функция, рассчитывающая размер буфера,
#  который необходим для закодированного потока
#
# \param[in] inp Список входных данных
# \return Размер буфера
def comp_enc_size(inp: list[int]) -> int:
    ostream_size = 2
    for elem in inp:
        match elem:
            case int(Defines.MESS_CODER_START_B) | \
                 int(Defines.MESS_CODER_END_B) | \
                 int(Defines.MESS_CODER_ENC_START):
                ostream_size += 2
            case _:
                ostream_size += 1
    return ostream_size


if __name__ == "__main__":
    print(f"{__file__}: It's not a program; it's a module")
